### PR TITLE
Honor MAX-ACCESS

### PR DESCRIPTION
**Pull request by derrell**
A provider now has a `maxAccess` member that limits what operations
are allowed. A `get` request requires at least "read-only" access, and
a `set` request requires at least "read-write".

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -4281,17 +4316,30 @@ Agent.prototype.request = function (requestMessage, rinfo) {
...
+						});
+					};
 			} else {
 				mibRequests[i] = new MibRequest ({
 					operation: requestPdu.type,
 					providerNode: providerNode,
 					instanceNode: instanceNode,
 					oid: requestPdu.varbinds[i].oid
 				});
-				if ( requestPdu.type == PduType.SetRequest ) {
```
Comment (markabrahams) [inline index.js:4286]: What is the reason for removing the PDU SetRequest type check?  The `setType` and `setValue` only pertain to SetRequests, so leaving these on getRequests is meaningless/clutter at best, if not deceptive.  As such, I'm inclined to keep this check.

**Comment by markabrahams** [comment]
This looks great thanks @derrell!  I would like to merge this, but before I do, I've just asked an inline question about removing the PDU SetRequest check - if you could clarify there, that would be great thank you.

File: index.js (hunk 1 above)
Comment (derrell) [reply index.js:4286]: I can't see any good reason to elide it. I expect I deleted it by accident. I'll re-insert that test.

**Comment by derrell** [comment]
Ok, I've re-inserted that code. Sorry 'bout that!

**Comment by markabrahams** [comment]
No problems @derrell !

I've gone to test this PR and have picked up a few things during the testing:

(1) SMIv2 uses MAX-ACCESS, but SMIv1 uses ACCESS.  A goal of the library is to support them both, so we'll probably need to look at reading the SMIv1 "ACCESS" clause into a provider's maxAccess value.  Ultimately whether ACCESS or MAX-ACCESS is read in from the MIB definition, I think we want to end up with a "maxAccess" provider field with the set of values you've given (as specified in the "Access" definition in RFC 2578 section 2).  Since the values for ACCESS and MAX-ACCESS are slightly different, there will need to be a mapping from ACCESS values to MAX-ACCESS values, which will be largely guided by RFC 3584 Section 2.1.1.

(2) maxAccess is implemented at a provider level, but also needs to be implemented at a table column level for tables.  For example, the current code doesn't respect the maxAccess field of a table column in the provider definition.  The current code uses the maxAccess value at the root table level for the access level of all table columns, which isn't the correct behaviour.

(3) By default, maxAccess isn't defined, and so nothing is accessible if you leave them out of the provider definition.  We have some options here:
(a) Leave it this way, but I don't think it's a desirable state for nothing to work by default.
(b) Elect a default, say of "read-only" (which I think is the most sensible, but a breaking change), or "read-create", which would not be a breaking change, but perhaps less desirable security-wise.
(c) Make maxAccess a mandatory field in each scalar provider (at the root level), and in each table column for a table provider.  (and maybe still set a default maxAccess at the root of a table provider of "not-accessible", as after the change in (2), this won't be checked any more).

I tend to favour (c), although it's a breaking change.  I can signal this fact with a major version bump as per semver.

I think these three things need to addressed before I merge the PR.  Let me know your thoughts on this stuff - thanks!

**Comment by derrell** [comment]
Hi @markabrahams . Sounds reasonable to have v1 work similarly, even though it's very old. I will work on this today.

Since you favor the breaking change, requiring maxAccess be present in the provider, I will implement it that way. (Is MAX-ACCESS -- and ACCESS, for that matter -- a required field in a MIB definition? If not, then we might still want a default to use when loading the MIB via a Store where there are automatically-created providers.)

Regarding the table columns, I'm still trying to grok how tables work in this library. I'd like to implement #144, too, which requires fully understanding the table implementation as well.

**Comment by derrell** [comment]
@markabrahams I think I've implemented everything you commented on, and it's tested with a v2 MIB. I don't have any v1 capability set up, so although v1 ACCESS is now mapped to v2 MAX-ACCESS, that mapping capability is completely untested.

**Comment by markabrahams** [comment]
Great - thanks @derrell !

Yes, the MAX-ACCESS clause is required for all OBJECT-TYPE MIB entries.  From RFC 2578 Section 7.3:
'The MAX-ACCESS clause, which must be present, defines whether it makes "protocol sense" to read, write and/or create an instance of the object, or to include its value in a notification.'

There are a few MIBs around that haven't transitioned from SMIv1 to SMIv2 (after 20 years!), hence the SMIv1 support requirement.

I've tested your updates with an SMIv1 MIB and have made a small but important change to the AccessToMaxAccess typing to get that working.

I've also changed agent.isAllowed() to get the provider type from the provider definition, and have fixed the column calculation to cater for composite indexes.

I've also had to fix the MIB parser to read the SMIv1 ACCESS clause, which it wasn't doing properly before (unrelated to your code).

I've released the modified PR to master now, and deployed as version 3.0.0 of the npm.  Major version bumped due to the breaking change requiring "maxAccess" in the scalar provider and table provider column definitions.

**Comment by derrell** [comment]
Thanks @markabrahams !

It does not appear that your change to `AccessToMaxAccess` is correct, though. In master, it maps every key to itself, i.e., it is an identity mapping. The purpose of that object is to map from a value of ACCESS from v1 to an equivalent value of MAX-ACCESS from v2. The keys, therefore,  should be the possible values of ACCESS, but you've changed them to be the possible values of MAX_ACCESS. At present, "none" will work as intended only because of the ternary at index.js:3132, which is intended as a fail-safe, not as something "expected" to be used. "none" should be one of the keys in `AccessToMaxAccess`. Furthermore, `AccessToMaxAccess` is intended to return the numeric MAX-ACCESS value, to assign to the `maxAccess` variable, but as currently written, `if typeof mibEntry["MAX-ACCESS"] != "undefined"` is true then `maxAccess` will be set to a numeric value, but if it is false then `maxAccess` will be set to a string. I don't think that's what we're looking for here.

Am I missing something?

**Comment by derrell** [comment]
Upon further reflection, the latter part of my statement above is incorrect. `maxAccess` does get a string value, which is later converted to its corresponding numeric value when assigned to the provider. However, the identity mapping of `AccessToMaxAccess` still appears incorrect, I think.

**Comment by markabrahams** [comment]
I made two changes to AccessToMaxAccess.

One was to change the values from integers to strings, to match your code (the other option would have been to change the code to handle the integer values).

The second was to change the AccessToMaxAccess keys to match the SMIv1 "ACCESS" values, given in RFC 1155 Section 4.3:
"read-only"
"read-write"
"write-only"
"not-accessible"

The only one that needed to change was "none" to "not-accessible" (as "none" isn't a valid SMIv1 "ACCESS" value).  And it turns out to be very close to an identity mapping - the only exception is SMIv1 "write-only", which doesn't have an SMIv2 MAX-ACCESS equivalent (and you had this correctly mapped to MAX-ACCESS "read-write" as per RFC 3584 Section 2.1.1).

**Comment by derrell** [comment]
Odd. My research yesterday discovered "none" was one of the options, but the RFC is pretty clear that it's not. I must not have been looking at an authoritative source. Disregard. All set. Thanks!

**Comment by markabrahams** [comment]
Excellent - this was a good feature to implement, so thanks again!  Will close PR and issue now.

**Comment by derrell** [comment]
Hmmm... snmpwalk now seems to abort early, with noAccess. Maybe something isn't quite right?

**Comment by markabrahams** [comment]
I can do a snmpwalk across the entire (albeit small!) MIB in examples/snmp-agent.js.

Starting the agent like:
mark@hope:~/node-net-snmp$ node example/snmp-agent.js -p 1161

and walking like:
mark@hope:~/node-net-snmp$ snmpwalk -v 2c -c public localhost:1161 1.3.6.1
SNMPv2-MIB::sysDescr.0 = STRING: Rage inside the machine!
IF-MIB::ifIndex.1 = INTEGER: 1
IF-MIB::ifIndex.2 = INTEGER: 2
IF-MIB::ifDescr.1 = STRING: lo
IF-MIB::ifDescr.2 = STRING: eth0
IF-MIB::ifType.1 = INTEGER: softwareLoopback(24)
IF-MIB::ifType.2 = INTEGER: myrinet(99)
IF-MIB::ifType.2 = No more variables left in this MIB View (It is past the end of the MIB tree)

succeeds.

Can you post a sample provider definition or MIB state where an snmpwalk fails?

**Comment by markabrahams** [comment]
Whoops - I re-opened the PR - was meant to re-open the issue (#145), which I've done now.  Will continue the discussion there.

**Pull request closed without merging**